src/internal/hash.c

```diff
@@ -0,0 +1,226 @@
+static inline int hash_version();
+static inline void hash_bytes(R_outpstream_t stream, void* p_input, int n);
+static inline void hash_char(R_outpstream_t stream, int input);
+
+sexp* rlang_hash(sexp* x) {
+  XXH3_state_t* p_xx_state = XXH3_createState();
+  XXH_errorcode err = XXH3_128bits_reset(p_xx_state);
+
+  if (err == XXH_ERROR) {
+    XXH3_freeState(p_xx_state);
+    r_abort("Couldn't initialize hash state.");
+  }
+
+  struct hash_state_t state = new_hash_state(p_xx_state);
+
+  int version = hash_version();
+
+  // Unused
+  sexp* (*hook)(sexp*, sexp*) = NULL;
+  sexp* hook_data = r_null;
+
+  struct R_outpstream_st stream;
+
+  R_InitOutPStream(
+    &stream,
+    (R_pstream_data_t) &state,
+    R_pstream_binary_format,
+    version,
+    hash_char,
+    hash_bytes,
+    hook,
+    hook_data
+  );
+
+  R_Serialize(x, &stream);
+
+  XXH128_hash_t hash = XXH3_128bits_digest(p_xx_state);
+  XXH3_freeState(p_xx_state);
+
+  // R assumes C99, so these are always defined as `uint64_t` in xxhash.h
+  XXH64_hash_t high = hash.high64;
+  XXH64_hash_t low = hash.low64;
+
+  // 32 for hash, 1 for terminating null added by `sprintf()`
+  char out[32 + 1];
+
+  sprintf(out, "%" PRIx64 "%" PRIx64, high, low);
+
+  return r_chr(out);
+}
+
+static inline
+struct hash_state_t new_hash_state(XXH3_state_t* p_xx_state) {
+  return (struct hash_state_t) {
+    .skip = true,
+    .n_skipped = 0,
+#if USE_VERSION_3
+    .n_native_enc = 0,
+#endif
+    .p_xx_state = p_xx_state
+  };
+}
+
+static inline
+int hash_version() {
+#if USE_VERSION_3
+  return 3;
+#else
+  return 2;
+#endif
+}
+
+static inline void hash_skip(struct hash_state_t* p_state, void* p_input, int n);
+
+static inline
+void hash_bytes(R_outpstream_t stream, void* p_input, int n) {
+  struct hash_state_t* p_state = (struct hash_state_t*) stream->data;
+
+  if (p_state->skip) {
+    hash_skip(p_state, p_input, n);
+    return;
+  }
+
+  XXH3_state_t* p_xx_state = p_state->p_xx_state;
+  XXH_errorcode err = XXH3_128bits_update(p_xx_state, p_input, n);
+
+  if (err == XXH_ERROR) {
+    XXH3_freeState(p_xx_state);
+    r_abort("Couldn't update hash state.");
+  }
+}
+
+static inline
+void hash_char(R_outpstream_t stream, int input) {
+  static int n = 1;
+  unsigned char byte = (unsigned char) input;
+  unsigned char* p_byte = &byte;
+  hash_bytes(stream, p_byte, n);
+}
+
+#if USE_VERSION_3
+
+static inline
+void hash_skip(struct hash_state_t* p_state, void* p_input, int n) {
+  if (p_state->n_skipped == N_BYTES_SERIALIZATION_INFO) {
+    // We've skipped all serialization info bytes.
+    // Incoming bytes tell the size of the native encoding string.
+    int* p_x = (int*) p_input;
+    p_state->n_native_enc = *p_x;
+
+    p_state->n_skipped += n;
+
+    return;
+  }
+
+  p_state->n_skipped += n;
+
+  // Haven't reached native encoding size yet,
+  // there is no way we have skipped everything
+  if (p_state->n_native_enc == 0) {
+    return;
+  }
```

Comment:
This might be easier to follow if this were the first branch with `p_state->n_skipped < N_BYTES_SERIALIZATION_INFO` as condition? Then the order in the code corresponds to the skip progress: skip until we reach the info's number of bytes, then record the native encoding, then skip the encoding string.